stan/math/prim/fun/gaus_interp.hpp

```diff
@@ -0,0 +1,196 @@
+inline double min_diff(int n, const std::vector<Tx>& xs) {
+  double dmin = value_of(xs[1]) - value_of(xs[0]);
+  for (int i = 1; i < n - 1; i++) {
+    if (value_of(xs[i + 1]) - value_of(xs[i]) < dmin) {
+      dmin = value_of(xs[i + 1]) - value_of(xs[i]);
+    }
+  }
+  return dmin;
+}
+
+}  // namespace internal
+
+/**
+ * Given a set of reference points \f$(xs_i, ys_i)\f$, create a mollifier
+ * that intersects the reference points. This function requires as input
+ * a struct created by the function gaus_interp_precomp. The algorithm
+ * used to create the mollifier is an iterative algorithm that works
+ * as follows. First a linear
+ * interpolation is created through the reference points. Then, the
+ * linear interpolation is convolved with a Gaussian whose width is
+ * proportional the smallest distance between successive points
+ * \f$xs_i\f$ and \f$xs_{i+1}\f$. Since the convolution is unlikely to
+ * intersect the reference points, the y-values of the reference points
+ * are shifted and the process repeats itself until the interpolation
+ * intersects all reference points.
+ *
+ * Note: This interpolation scheme should be used when the function
+ * to be interpolated is well-resolved by the reference points.
+ *
+ * @param xs vector of independent variable of reference points
+ * @param ys vector of dependent variable of reference points
+ * @param params a vector created by gaus_interp_precomp
+ * @param x the point at which to evaluate the interpolation
+ * @return value of the interpolation at x
+ */
+template <typename Tx>
+inline return_type_t<Tx> gaus_interp(const std::vector<double>& xs,
+                                     const std::vector<double>& ys,
+                                     const std::vector<double>& params,
+                                     const Tx& x) {
+  // enforce that interpolation point is between smallest and largest
+  // reference point
+  static char const* function = "gaus_interp";
+  check_less_or_equal(function, "Interpolation point", x, xs.back());
+  check_greater_or_equal(function, "Interpolation point", x, xs.front());
+  check_ordered(function, "xs", xs);
+  check_not_nan(function, "xs", xs);
+  check_not_nan(function, "ys", ys);
+  check_not_nan(function, "x", x);
+  check_greater(function, "xs", xs.size(), 1);
+
+  // number of standard deviations to extend endpoints for convolution
+  const double NSTDS = 10;
+  int n = xs.size();
+
+  // params is vector of the form (as, bs, sig2)
+
+  // create copy of xs so that endpoints can be extended
+  std::vector<double> xs2 = xs;
+
+  // extend out first and last lines for convolution
+  double sig = std::sqrt(params[2 * n - 2]);
+  xs2[0] = xs[0] - NSTDS * sig;
+  xs2[n - 1] = xs[n - 1] + NSTDS * sig;
+
+  // no need to convolve far from center of gaussian, so
+  // get lower and upper indexes for integration bounds
+  auto lb = std::lower_bound(xs.begin(), xs.end(), x - NSTDS * sig);
+  int ind_start = &(*lb) - &xs[0] - 1;
```

Comment:
Can you use `std::distance()` here to get the distance from the return of `lower_bound()` to the start of the vector?

Reply:
I couldn't remember if we settled on std::distance or pointer arithmetic in our call. Will change.